piq/iw_ssim.py

```diff
@@ -0,0 +1,458 @@
+    vv = sigma_xx - g * sigma_xy
+    g = g.masked_fill(sigma_yy < EPS, 0)
+    vv[sigma_yy < EPS] = sigma_xx[sigma_yy < EPS]
+    g = g.masked_fill(sigma_xx < EPS, 0)
+    vv = vv.masked_fill(sigma_xx < EPS, 0)
+
+    block = [kernel_size, kernel_size]
+
+    nblv = y.size(-2) - block[0] + 1
+    nblh = y.size(-1) - block[1] + 1
+    nexp = nblv * nblh
+    N = block[0] * block[1]
+
+    assert block[0] % 2 == 1 and block[1] % 2 == 1, f'Expected odd block dimensions, got {block}'
+
+    Ly = (block[0] - 1) // 2
+    Lx = (block[1] - 1) // 2
+
+    if y_parent is not None:
+        # upscale y_parent and cut to the size of y
+
+        y_parent_up = _image_enlarge(y_parent)[:, :, :y.size(-2), :y.size(-1)]
+        N = N + 1
+
+    Y = torch.zeros(y.size(0), y.size(1), nexp, N)
+
+    n = -1
+    for ny in range(-Ly, Ly + 1):
+        for nx in range(-Lx, Lx + 1):
+            n = n + 1
+            foo = _shift(y, [ny, nx])
+            foo = foo[:, :, Ly:Ly + nblv, Lx:Lx + nblh]
+            Y[..., n] = foo.flatten(start_dim=-2, end_dim=-1)
+
+    if y_parent is not None:
+        n = n + 1
+        foo = y_parent_up
+        foo = foo[:, :, Ly:Ly + nblv, Lx:Lx + nblh]
+        Y[..., n] = foo.flatten(start_dim=-2, end_dim=-1)
+
+    C_u = torch.matmul(Y.transpose(-2, -1), Y) / nexp
+
+    recommended_torch_version = _parse_version('1.7.0')
+    torch_version = _parse_version(torch.__version__)
+    if len(torch_version) != 0 and torch_version >= recommended_torch_version:
+        eig_values, eig_vectors = torch.linalg.eigh(C_u)
+    else:
+        eig_values, eig_vectors = torch.symeig(C_u, eigenvectors=True)
+
+    sum_eig_values = torch.sum(eig_values, dim=-1).view(y.size(0), y.size(1), 1, 1)
+    non_zero_eig_values_matrix = torch.diag_embed(eig_values * (eig_values > 0))
+    sum_non_zero_eig_values = torch.sum(non_zero_eig_values_matrix, dim=(-2, -1), keepdim=True)
+
+    L = non_zero_eig_values_matrix * sum_eig_values / (sum_non_zero_eig_values + (sum_non_zero_eig_values == 0))
+
+    C_u = torch.matmul(torch.matmul(eig_vectors, L), eig_vectors.transpose(-2, -1))
+
+    C_u_inv = torch.inverse(C_u)
+
+    ss = torch.matmul(Y, C_u_inv) * Y / N
+    ss = torch.sum(ss, dim=-1, keepdim=True)
+    ss = ss.view(y.size(0), y.size(1), nblv, nblh)
+    g = g[:, :, Ly: Ly + nblv, Lx: Lx + nblh]
+    vv = vv[:, :, Ly: Ly + nblv, Lx: Lx + nblh]
+
+    # Calculate mutual information
+    scaled_eig_values = torch.diagonal(L, offset=0, dim1=-2, dim2=-1).unsqueeze(2).unsqueeze(3)
+
+    iw_map = torch.sum(torch.log2(1 + ((vv.unsqueeze(-1) + (1 + g.unsqueeze(-1) * g.unsqueeze(-1)) * sigma_nsq)
+                                       * ss.unsqueeze(-1) * scaled_eig_values + sigma_nsq * vv.unsqueeze(-1)) / (
+                                              sigma_nsq * sigma_nsq)), dim=-1)
+
+    iw_map[iw_map < EPS] = 0
+
+    return iw_map
+
+
+def _image_enlarge(x: torch.Tensor) -> torch.Tensor:
+    r"""Bilinear upscaling of the image.
+
```

Comment:
This function does more than just image upscaling. Can you clarify what's happening in lines 436-440 in docstring?